---
title:  "<big><u>IMAP-WORKFLOWS</u></big><br><br>Snakemake workflows for microbiome and multi-omics data analysis<br><br>"
subtitle: |
  | ![](images/cicd.png)<br>
  | Fostering Reproducible Microbiome data Analysis with Snakemake workflow
date:
- MAP-WORKFLOWS<br>[Latest GitHub-Repo](https://github.com/tmbuza/imap-snakemake-workflows/)<br>
- Maintained by Teresia Mrema-Buza<br><br>
- <b>Updated on `r Sys.Date()`</b>
author: Teresia Mrema-Buza
output: 
  html_document: 
css:
- styles.css
csl: "library/apa.csl"
bibliography:
- library/references.bib
- library/imap.bib
link-citations: true
mainfont:
biblio-style: apalike
email_address: "tmbuza@complexdatainsights.com"
github_repo: "https://github.com/tmbuza/imap-snakemake-workflows"
description: |
  | How to get started with systematic microbiome data analysis with Snakemake workflows.
---


<div id="main-container">

<link rel="preconnect" href="https://fonts.googleapis.com">
<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
<link href="https://fonts.googleapis.com/css2?family=Anton" rel="stylesheet">
<link href="https://fonts.googleapis.com/css2?family=Roboto:wght@100;300;400;500;700,900&display=swap" rel="stylesheet">
<link href="https://fonts.googleapis.com/css2?family=Oswald:wght@300;400;700&display=swap" rel="stylesheet">
<link href="https://fonts.googleapis.com/css2?family=Merriweather:wght@300;400;700&display=swap" rel="stylesheet">
<link href="https://fonts.googleapis.com/css2?family=Montserrat:wght@100;200;300;400;700&display=swap" rel="stylesheet">

```{r globalsetup, echo=FALSE, message=FALSE, warning=FALSE}
knitr::opts_chunk$set(
  echo  =FALSE,
  message  =FALSE,
  warning  =FALSE,
  cache  =FALSE,
  comment  =NA,
  fig.path='./figures/',
  fig.show='asis',
  dev  ='png',
  fig.align='center',
  out.width  ="70%",
  fig.width  =7,
  fig.asp  =0.7,
  fig.show  ="asis"
)

library(tidyverse, suppressPackageStartupMessages())
library(schtools)
```

# General Overview of IMAP project

**IMAP** stands for **Integrated Microbiome Analysis Pipelines** but can also stands for **Integrated Multi-Omics Analysis Pipelines**. 
IMAP for microbiome analysis comprises different parts which individually, represents a standalone GitHub repository prefixed with {**imap-**}. When used sequentially, these parts can provide a systematic microbiome data analysis beyond the traditional methods. 

<br>

![IMAP tentative parts: Each part forms a standalone git repository containing similar project stucture.](images/imap.png)

<br>


## IMAP approach
- We use the snakemake workflow management system [@Koster2021; @Snakemake2023] to:
  - Maintain reproducibility in technical validation and regeneration of results.
  - Create scalable data analysis.
  - Foster sustainable improvement of the microbiome data analysis.
  - Ensure adaptability that enables users to modify the analysis to answer extended research questions.
  - Ensure transparency of the technical analysis methodologies.
- We break complex workflows into small contiguous but related chunks where each major step forms a separate executable snakemake rule.

<br>

## Mission and Vision
We envision fostering continuous integration and improvement of highly reproducible and sustainable workflows for microbiome data analysis.

<br>

## IMAP snakemake workflow series
The workflows define the **snakemake rules** in a DAG (directed acyclic graph) format. 
You may find a link to the interactive snakemake HTML report in some of the workflows. The link will enable you explore the analysis workflow, associated statistics and in some the result. While exploring the interactive workflow, you can close the left bar (the overlap) to get a more expansive display view. 

> NOTE! Use a wider screen (landscape) when viewing the interactive report. 

<br>

# Category 01: Microbiome snakemake workflows

## IMAP-PART 01 
### Title: Software requirement for microbiome data analysis with Snakemake workflows
 
<br>

![Explore the IMAP-PART 01 interactive snakemake report (if available) [here.](https://complexdatainsights.com/imap/imap-software-requirement/report.html)](images/imap_part01.svg)

<br>

## IMAP-PART 02 
### Title: Downloading and exploring microbiome sample metadata from SRA Database

<br>

![Explore the IMAP-PART 02 interactive snakemake report (if available) [here.](https://complexdatainsights.com/imap/imap-sample-metadata/report.html)](images/imap_part02.svg)

<br>

## IMAP-PART 03 
### Title: Downloading microbiome fastq sequences from SRA database

<br>

![Explore the IMAP-PART 03 interactive snakemake report (if available) [here.](https://complexdatainsights.com/imap/imap-download-sra-reads/report.html)](images/imap_part03.svg)

<br>

## IMAP-PART 04 
### Title: Quality Control of Microbiome Next Generation Sequencing Reads

<br>

![Explore the IMAP-PART 04 interactive snakemake report (if available) [here.](https://complexdatainsights.com/imap/imap-read-quality-control/report.html)](images/imap_part04.svg)

<br>

## IMAP-PART 05 
### Title: Microbial profiling using Mothur pipeline with Snakemake workflows

<br>

![Explore the IMAP-PART 05 interactive snakemake report (if available) [here.](https://complexdatainsights.com/imap/imap-bioinformatics-mothur/report.html)](images/imap_part05.svg)

<br>

## IMAP-PART 06
### Title: Microbial profiling using QIIME2 pipeline with Snakemake workflows

<br>

![Explore the IMAP-PART 06 interactive snakemake report (if available) [here.](https://complexdatainsights.com/imap/imap-bioinformatics-qiime2/report.html)](images/imap_part06.svg)

<br>

## IMAP-PART 07
### Title: Processing Output from 16S-based microbiome bioinformatics pipelines

<br>

![Explore the IMAP-PART 07 interactive snakemake report (if available) [here.](https://complexdatainsights.com/imap/imap-data-processing/report.html)](images/imap_part07.svg)


## IMAP-PART 08
### Title: Exploratory Analysis of 16S-Based Microbial Profiles

<br>

![Explore the IMAP-PART 08 interactive snakemake report (if available) [here.](https://complexdatainsights.com/imap/imap-exploratory-analysis/report.html)](images/imap_part08.svg)



## IMAP-PART 09
### Title: Multi-Dimensional Data Visualization with integrated tools and Snakemake workflow

<br>

![Explore the IMAP-PART 09 interactive snakemake report (if available) [here.](https://complexdatainsights.com/imap/imap-visualize-abundance/report.html)](images/imap_part09.svg)



## IMAP-PART 10
### Title: Microbiome Alpha diversity analysis

<br>

![Explore the IMAP-PART 10 interactive snakemake report (if available) [here.](https://complexdatainsights.com/imap/#/report.html)](images/imap_part10.svg)


## IMAP-PART 11
### Title: Microbiome Beta diversity analysis

<br>

![Explore the IMAP-PART 11 interactive snakemake report (if available) [here.](https://complexdatainsights.com/imap/#/report.html)](images/imap_part11.svg)


<br><hr width=100%><br>

## Related work
```{block imaprepos, echo=TRUE}

| Repo | Description | Status |
|:-----------------------|----------------------------------|--------------:|
| [IMAP-OVERVIEW](https://github.com/datainsights/imap-project-overview/) | IMAP project overview | [In-progress](https://datainsights.github.io/imap-project-overview/) |
| [IMAP-PART 01](https://github.com/tmbuza/imap-software-requirements/) | Software requirement for microbiome data analysis with Snakemake workflows | [In-progress](https://tmbuza.github.io/imap-software-requirements/) |
| [IMAP-PART 02](https://github.com/tmbuza/imap-sample-metadata/) | Downloading and exploring microbiome sample metadata from SRA Database | [In-progress](https://tmbuza.github.io/imap-sample-metadata/) |
| [IMAP-PART 03](https://github.com/tmbuza/imap-download-sra-reads/) | Downloading and filtering microbiome sequencing data from SRA database | [In-progress](https://tmbuza.github.io/imap-download-sra-reads/) |
| [IMAP-PART 04](https://github.com/tmbuza/imap-read-quality-control/) | Quality Control of Microbiome Next Generation Sequencing Reads | [In-progress](https://tmbuza.github.io/imap-read-quality-control/) |
| [IMAP-PART 05](https://github.com/tmbuza/imap-bioinformatics-mothur/) | Microbial profiling using MOTHUR and Snakemake workflows | [In-progress](https://tmbuza.github.io/imap-bioinformatics-mothur/) |
| [IMAP-PART 06](https://github.com/tmbuza/imap-bioinformatics-qiime2/) | Microbial profiling using QIIME2 and Snakemake workflows | [In-progress](https://tmbuza.github.io/imap-bioinformatics-qiime2/) |
| [IMAP-PART 07](https://github.com/tmbuza/imap-data-processing/) | Processing Output from 16S-Based microbiome bioinformatics pipelines| [In-progress](https://tmbuza.github.io/imap-data-processing/) |
| [IMAP-PART 08](https://github.com/tmbuza/imap-exploratory-analysis/) | Exploratory Analysis of 16S-Based Microbiome Processed Data | [In-progress](https://tmbuza.github.io/imap-exploratory-analysis/) |
| [IMAP-SUMMARY](https://github.com/tmbuza/imap-snakemake-workflows/) | Summary of snakemake workflows for microbiome data analysis | [In-progress](https://tmbuza.github.io/imap-snakemake-workflows/) |
```

<br><hr width=100%><hr width=100%><br>

# Multi-Omics data analysis

...BRIEF INTRO IN PROGRESS...

<br>

## Citation

> Please consider citing the iMAP article [@TMBuza2019] if you find any part of the IMAP practical user guides helpful in your microbiome data analysis.

<br>

# References
::: {#refs}
:::

<br><hr width=100%><br>

# Appendix {-}

## Project main tree
```{bash, include=FALSE}
bash workflow/scripts/tree.sh
```

```{bash treeimg}
cat results/project_tree.txt
```

<br>

![](images/smkreport/screenshot.png)

<br>


## Troubleshooting of FAQs
<ol>
  <li>Question</li>
    <ul>Answer</ul>
  <li>Question</li>
    <ul>Answer</ul>
</ol>


</div> <!-- end main-container -->

<br><hr width=100%><br>

<div id="footer">
Last updated on `r format(Sys.time(), '%B %d, %Y')`.<br>
Github_Repo: `r rmarkdown::metadata$github_repo`.<br>
Repo compiled and maintained by: [`r rmarkdown::metadata$author`](mailto:`r rmarkdown::metadata$email_address`). 
</div>
